### DISCUSSION TITLE
Yaml: Help setting json schemas

### DISCUSSION THREAD

**Discussion by mikeymop** [zed-industries/zed#14528] [Feature Requests]
Hello,

I've been experimenting with Zed on some of my work projects, and noticed that the ability to select a Yaml schema is not exposed in the bottom toolbar when `YAML` is selected.

I saw in the Zed documentation that we can customize the [lsp settings](https://zed.dev/docs/configuring-zed#lsp) for a given lsp.

For [yaml](https://zed.dev/docs/languages/yaml) I saw that we have a few options for getting, and setting json schemas for a given project.

The most useful option would be to set a [glob pattern](https://github.com/redhat-developer/yaml-language-server?tab=readme-ov-file#using-yamlschemas-settings) using a local zed `settings.json` such as in `.zed/settings.json` in a project. We have `.vscode` and `.idea` local settings in most of our repositories.

Following this I created a helm chart with `helm create zed-example` and set the following in `.zed/settings.json`:
```json
{
  "lsp": {
    "yaml-language-server": {
      "initialization_options": {
        "yaml": {
          "schemas": {
            "http://json.schemastore.org/chart.json": "templates/*.yaml"
          }
        }
      }
    }
  }
}
```

This does not enable me set the yaml schema in the template example such as with `templates/deployment.yaml`

I have also tried

```json
{
  "lsp": {
    "yaml-language-server": {
      "initialization_options": {
        "yaml.schemas": {
          "http://json.schemastore.org/chart.json": "templates/*.yaml"
        }
      }
    }
  }
}
```

With no luck.

I can tell this is using `yaml-language-server` and is working because if I set the yaml schema in-line it works:

![Screenshot from 2024-07-15 19-24-16](https://github.com/user-attachments/assets/bab0ad32-c8c7-4212-9a3a-9d37a8d94c0b)

I have also tried this in the global `settings.json`, and outside of `initialization_options` (eg: `lsp.yaml-language-server.yaml.schemas` ) with no luck.

Is there another way to set yaml schemas, per-project, using relative file glob patterns?

**Comment by gmeligio** [comment]
I'm experiencing the same issue. I noticed in the RPC message logs that only the inline schema is being searched. When it's configured in `settings.json`, the LSP doesn't try to find the schema. This might help you to debug if the configuration is not behaving as expected.

Zed 0.151.2
Kubuntu 24.04
![image](https://github.com/user-attachments/assets/a3531445-5e61-4577-97a2-f9f49a6a32e9)

**Reply by gmeligio** [reply]
Adding this issue for context https://github.com/zed-industries/zed/issues/7636

**Comment by notpeter** [comment]
I haven't messed with it too much, but we should support specifying the schema either via the inline modeline (which sounds like it's working for you) or via the LSP settings (which sounds like it is not working).  I would expect if you have both that the inline would override the one provided by settings.

Looking the [yaml-language-server docs](https://github.com/redhat-developer/yaml-language-server?tab=readme-ov-file#more-examples-of-schema-association), nearly all of their examples include a leading slash on the glob pattern.  Could you try and see if instead of specifying `["templates/*.yml"]` explicitly setting `["/templates/*.yml"]` or `["**/templates/*.yml"]` yields better results. If so, we can update the docs.

**Comment by softprops** [comment]
something I noticed what was [mentioned above](https://github.com/zed-industries/zed/discussions/14528#discussioncomment-10593930) and was seeing UI feedback only after adding the inline schema mentioned in [zed's docs](https://zed.dev/docs/languages/yaml#schemas) until I switched from using the `settings` key to the `initialization_options`. Afterwards I noticed the yaml language server picked up the configuration. This might just be a bug with zed's docs.

However I noticed additional unexpected behavior that might be related to zeds deserialization of settings before passing it off the the lsp. 

In my configuration I have

```json
 "lsp": {
    "yaml-language-server": {
      "initialization_options": {
        "yaml": {
          "schemas": {
            "feature-flags/schema.json": ["feature-flags/definitions/**/*.yml"]
          }
        }
      }
    }
  }
```

Expected behavior: This should read the schema from from `feature-flags/schema.json` for any file that matches `feature-flags/definitions/**/*.yml`. 

Actual behavior: The file match does work but zed's ui highlights an error in any opened matching doc saying `YAML: Unable to load schema from '/path/to/project/root/schema.json': No content.`. It seems to be missing the `feature-flags/` path of the declared path. I've also tried `/feature-flags/schema.json` which gets interpretted as an absolute path  as well as `./feature-flags/schema.json` which comically gets translated to `YAML: Unable to load schema from `/path/to/project/root/feature-flags,schema.json': No content.` Note the path slash gets translated to a comma!?

What did work was `/path/to/project/root/feature-flags/schema.json` as the schemas setting which is less than ideal because it means that my configuration is no longer portable. It's hardcoded to a directory subject to change.

**Reply by gmeligio** [reply]
Thank you. This fixed my issue after replacing `settings` with `initialization_options`. So, it's a documentation issue.

My perception is that the LSP settings started working after https://github.com/zed-industries/zed/pull/17479 but it introduced this documentation issue. It would be great if a PR is submitted to fix the documentation. I'll try to get some time to contribute myself.

This is in Zed 0.153.6.

-----

Regarding using a local relative path, that could be a requested as a new feature maybe. As a workaround for having a portable path, you could upload the secham to a Github gist or somewhere public.

**Reply by notpeter** [reply]
> My perception is that the LSP settings started working after https://github.com/zed-industries/zed/pull/17479 but it introduced this documentation issue. It would be great if a PR is submitted to fix the documentation. I'll try to get some time to contribute myself.

The issue is that there was an additional change which moved yaml-server-settings from `initialization_options` (incorrect) to `settings` (correct) under lsp.

The docs now include the correct example.  Use `lsp.yaml-language-server.settings` not `lsp.yaml-lanage-server.initialization_options`

https://zed.dev/docs/languages/yaml

**Reply by gmeligio** [reply]
Thanks. I can confirm that this is working OK with `lsp.yaml-lanage-server.settings` in Zed 0.154.1. It was working with `lsp.yaml-lanage-server.initialization_options` before that version.

I didn't debug much about why. I tried reinstalling multiple times and cleaning settings before 0.154.1, thinking that was in my OS, but it was Zed.

**Reply by notpeter** [reply]
We moved from `initialization_options` to workspace `settings` because the way we were handling it was incorrect, we were sending the same block of `initialization_options` to both `initialization_options` on startup and on workspace settings when they changed.

We now handle them separately because they are in fact separate things as far as the language server in concerned. Apologies for the lack of documentation about this change but I've since updated the docs to reflect the way to use it going forward.

**Reply by gmeligio** [reply]
Thanks for the clarification :+1: .

**Reply by softprops** [reply]
I tried settings first because that’s what’s in the docs. As mentioned earlier that didn’t do anything. I also checked to make sure to was on the most recent release version of zed at the time.

**Reply by dbedggood** [reply]
I'm facing the same issue

**Comment by notpeter** [comment]
We should definitely try amd support project-relative pathing for schema json.

Also curious whether vscode with yaml-language-server works with the relative paths you've tried.  If they don't work there, then it's an upstream yaml-language-server issue. If it works in vscode but not zed please file an issue (ideally with a minimal reproduction repo showing it working in vscode but broken in Zed).

**Reply by mikeymop** [reply]
I still have the test repo I used for the bug report handy.

When I get home from work I can reply back here with it

**Reply by mikeymop** [reply]
Here is a [reproduction](https://github.com/mikeymop/zed-jsonschema-repro) using `helm create`.

**Reply by notpeter** [reply]
A few things.
1. You're repo uses initialization_options not `settings`.
2. The `Chart.yaml` works (for me) no settings required.
3. You're attempting to associate the Chart schema with files that are not charts.  If you look at the chart.json there is no `kind`.  This is the wrong schema and `yaml-language-server` is telling you that.
4. Helm charts aren't YAML, they are YAML with Go template escaping.  These do not work out of the box with `yaml-language-server` 
<img width="716" alt="Screenshot 2024-09-24 at 1 17 05" src="https://github.com/user-attachments/assets/fb996243-de95-4914-8b24-7290a1e4ad22">

Are you seeing something different?

This isn't a Zed settings, yaml-language-server configuration or JSON schema issue at all.
You can't validate the JSON schema of a template.
It's like complaining that a PHP script doesn't validate as html:
```php
<html><body><<?php echo '<p>Hello World</p>'; ?></body></html>
```

The features you are looking for, a zed extension which integrated helm-ls and an as of yet unwritten go-template+yaml tree-sitter, does not yet exist.  But there is an issue: 
- https://github.com/zed-industries/extensions/issues/186

**Reply by softprops** [reply]
Re relative paths. Relative paths work in both vscode and IntelliJ but I’m not sure if IntelliJ uses the yaml language server. They do not work in zed.

**Reply by mikeymop** [reply]
I understand it's not effective to use jsonschemas for template yamls inside a helm chart. I only used that as an example.

Something that is useful is to define a `values.schema.json` for the chart values which can be enforced using `helm lint`.

I am able to get this working on `0.186.9` using `lsp.yaml-language-server.settings.yaml.shemas` eg:
```json
{
  "lsp": {
    "yaml-language-server": {
      "settings": {
        "yaml": {
          "schemas": {
            "https://example.com/config.schema.json": [
              "example/config/config.yaml",
            ]
          }
        }
      }
    }
  }
}
```

The updated [documentation page](https://zed.dev/docs/languages/yaml?highlight=yaml#configuration) does point out that local schemas can be specified. (Thank you [notpeter](https://github.com/zed-industries/zed/discussions/14528#discussioncomment-10765050))

However with the above example, if I saved a file in `example/config/schemas/config.schema.json` and instead used:

```diff
{
  "lsp": {
    "yaml-language-server": {
      "settings": {
        "yaml": {
          "schemas": {
-            "https://example.com/config.schema.json": [
+            "./example/config/schemas/config.schema.json": [
              "example/config/config.yaml",
            ]
          }
        }
      }
    }
  }
}
```
I get the error:
> Unable to load schema from '/home/mikey/dev/project/example,config,schema,v1.9,config.schema.json': No content. (YAML 768)

Which is consistent with [softprops](https://github.com/zed-industries/zed/discussions/14528#discussioncomment-10721511) observation.

It seems the parser is assuming that the schemas will always be at the top-level of the project rather than parsing out the entire relative path.

Do I have to keep the local schema at the root of my project?
If it helps, I can update the reproduction as a reference.

**Comment by notpeter** [comment]
>  I've also tried /feature-flags/schema.json which gets interpretted as an absolute path as well as ./feature-flags/schema.json which comically gets translated to YAML: Unable to load schema from /path/to/project/root/feature-flags,schema.json': No content.` Note the path slash gets translated to a comma!?

That's super wonky.  Looking at the `yaml-language-server` source it failed to parse the json schema:
https://github.com/redhat-developer/yaml-language-server/blob/f039273ee5b6974db5ad34b03bb24cfe09b64447/src/languageservice/services/yamlSchemaService.ts#L636-L648

As for the comma, do you happen to live in a country which has commas for decimal separator? (Germany, etc) I wonder whether that call to `localize` is a bit misguided.

Anyways, if anyone could provide me with a self-contained reproduction repo to start with I might be able to find some time to track this down.

**Reply by softprops** [reply]
I live in the us

**Comment by kevinvalk** [comment]
@notpeter  it took some time clearly, but here you go. I created a full reproducible example repository that can be opened in both VSCode and Zed to illustrate the difference directly. It illustrates both this YAML bug, but also a JSON bug in JSON schemas (they seem unrelated).

See https://github.com/kevinvalk/issue-zed-yaml-json-schema

Somebody also already opened this issue at https://github.com/zed-industries/zed/issues/30938

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
